target_chains/solana/sdk/js/pyth_solana_receiver/src/PythSolanaReceiver.ts

```diff
@@ -39,12 +39,150 @@ import {
   PriorityFeeConfig,
 } from "@pythnetwork/solana-utils";
 
+/**
+ * A builder class to build transactions that:
+ * - Post price updates (fully or partially verified)
+ * - Consume price updates in a consumer program
+ * - (Optionally) Close price update and encoded vaa accounts to recover the rent
+ *
+ * @example
+ * ```typescript
+ *  const priceUpdateData = await priceServiceConnection.getLatestVaas([
+ *    SOL_PRICE_FEED_ID,
+ *    ETH_PRICE_FEED_ID,
+ *  ]);
+ *
+ * const transactionBuilder = pythSolanaReceiver.newTransactionBuilder();
+ * await transactionBuilder.withPostPriceUpdates(priceUpdateData);
+ * await transactionBuilder.withPriceConsumerInstructions(...)
+ * transactionBuilder.withCloseInstructions();
+ *
+ * await pythSolanaReceiver.provider.sendAll(await transactionBuilder.getVersionedTransactions({computeUnitPriceMicroLamports:1000000}))
+ * ```
+ */
+export class PythTransactionBuilder extends TransactionBuilder {
+  readonly pythSolanaReceiver: PythSolanaReceiver;
+  private closeInstructions: InstructionWithEphemeralSigners[];
+  private priceFeedIdToPriceUpdateAccount: Record<string, PublicKey>;
+
+  constructor(pythSolanaReceiver: PythSolanaReceiver) {
+    super(pythSolanaReceiver.wallet.publicKey, pythSolanaReceiver.connection);
+    this.pythSolanaReceiver = pythSolanaReceiver;
+    this.closeInstructions = [];
+    this.priceFeedIdToPriceUpdateAccount = {};
+  }
+
+  /**
+   * Add instructions to post price updates to the builder.
+   *
+   * @param priceUpdateDataArray the output of the `@pythnetwork/price-service-client`'s `PriceServiceConnection.getLatestVaas`. This is an array of verifiable price updates.
+   */
+  async withPostPriceUpdates(priceUpdateDataArray: string[]) {
```

Comment:
Should we enforce that 
`withPostPriceUpdates`, `withPriceConsumerInstructions` and `withCloseInstructions` need to be called sequentially?

Reply:
I think it would be nice to expose them.

I thought there should be a neat way to do it with types. Look at [here](https://stackoverflow.com/questions/73072541/call-chain-order-dependent-builder-pattern-in-typescript-compiler-limitation) or [here](https://stackoverflow.com/questions/45291644/builder-pattern-using-typescript-interfaces) to see how you can impose order using some type markers. Feel free to ignore it if you think it gets complex.

Reply:
some of constraints here aren't strictly necessary right? Like in theory I could call `withPriceConsumerInstructions` multiple times adding different sets of instructions and the code will work (?)

My inclination would be to avoid adding constraints on usage that aren't necessary for correctness -- you're just making the builder less flexible. see also the comment below about `withCloseInstructions` which I think helps resolve the one correctness constraint that exists.